### PR TITLE
implemented Category, CategoryService, CategoryController

### PR DESCRIPTION
**Pull request by mmosii**

### CODE REVIEW

**Review by VolodymyrSkvira — APPROVED** [review]
LGTM

File: src/main/java/mmosii/bookstore/controller/BookController.java [hunk 1]
```diff
@@ -33,45 +33,45 @@
 public class BookController {
     private final BookService bookService;
 
-    @PreAuthorize("hasRole('USER')")
+    @PreAuthorize("hasAuthority('USER')")
```
Comment (NazarSmith228) [inline src/main/java/mmosii/bookstore/controller/BookController.java:37]: what is the reason of this change?

as far as I remember `hasRole` automatically adds `ROLE_` while `hasAuthority` does not, which means:
```
hasRole("USER) == hasAuthority("ROLE_USER")
```

so you decided to shorten the role names and therefore use `hasAuthority`?

File: src/main/java/mmosii/bookstore/dto/book/CreateBookRequestDto.java [hunk 2]
```diff
@@ -4,11 +4,13 @@
...
-                                  @NotBlank String isbn,
-                                  @NotNull @Positive BigDecimal price,
-                                  String description,
-                                  String coverImage) {
+                                   @NotBlank String author,
+                                   @NotBlank String isbn,
+                                   @NotNull @Positive BigDecimal price,
+                                   String description,
+                                   String coverImage,
+                                   List<Long> categoryIds) {
```
Comment (NazarSmith228) [inline src/main/java/mmosii/bookstore/dto/book/CreateBookRequestDto.java]: you can also validate categories:
```
@NotEmpty Set<@Positive Long> categoryIds
```

File: src/main/java/mmosii/bookstore/repository/book/BookRepository.java [hunk 3]
```diff
@@ -1,8 +1,12 @@
 package mmosii.bookstore.repository.book;
 
+import java.util.List;
 import mmosii.bookstore.model.Book;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
 
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
```
Comment (NazarSmith228) [inline src/main/java/mmosii/bookstore/repository/book/BookRepository.java:11]: `@Repository` is missing

File: src/main/java/mmosii/bookstore/repository/book/CategoryRepository.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package mmosii.bookstore.repository.book;
+
+import mmosii.bookstore.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
```
Comment (NazarSmith228) [inline src/main/java/mmosii/bookstore/repository/book/CategoryRepository.java:8]: `@Repository` is missing

File: src/main/java/mmosii/bookstore/service/impl/BookServiceImpl.java [hunk 5]
```diff
@@ -63,9 +70,23 @@ public List<BookDto> search(BookSearchParametersDto searchParameters) {
...
+    }
+
     private void checkIfBookExistsById(Long id) {
         if (!bookRepository.existsById(id)) {
             throw new EntityNotFoundException("Can't find book with id " + id);
         }
     }
+
+    private void setCategories(CreateBookRequestDto requestDto, Book book) {
+        if (requestDto.categoryIds() != null) {
```
Comment (NazarSmith228) [inline src/main/java/mmosii/bookstore/service/impl/BookServiceImpl.java]: this check should be performed via validation

File: src/main/java/mmosii/bookstore/service/impl/BookServiceImpl.java [hunk 6]
```diff
@@ -63,9 +70,23 @@ public List<BookDto> search(BookSearchParametersDto searchParameters) {
...
+            book.setCategories(new HashSet<>(
+                    categoryRepository.findAllById(requestDto.categoryIds())));
```
Comment (NazarSmith228) [inline src/main/java/mmosii/bookstore/service/impl/BookServiceImpl.java]: I would suggest using `getReference` since you don't need to load the caregories here - you just need to create a mapping

File: src/main/java/mmosii/bookstore/service/impl/CategoryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    @Override
+    public CategoryDto save(CreateCategoryRequestDto categoryDto) {
+        return categoryMapper.toDto(categoryRepository.save(categoryMapper.toEntity(categoryDto)));
+    }
+
+    @Override
+    public CategoryDto update(Long id, CreateCategoryRequestDto categoryDto) {
+        checkIfCategoryExistsById(id);
+        Category category = categoryMapper.toEntity(categoryDto);
+        category.setId(id);
```
Comment (NazarSmith228) [inline src/main/java/mmosii/bookstore/service/impl/CategoryServiceImpl.java]: this is a bit wrong

you should get the category by id, update it (via Mapstruct) and then save

File: src/main/java/mmosii/bookstore/service/impl/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    public CategoryDto getById(Long id) {
+        return categoryRepository
+                .findById(id)
+                .map(categoryMapper::toDto)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find category by id " + id));
+    }
+
+    @Override
+    public CategoryDto save(CreateCategoryRequestDto categoryDto) {
+        return categoryMapper.toDto(categoryRepository.save(categoryMapper.toEntity(categoryDto)));
```
Comment (NazarSmith228) [inline src/main/java/mmosii/bookstore/service/impl/CategoryServiceImpl.java]: I would define a separate variable for `toEntity` - this would improve readability

File: src/main/resources/db/changelog/changes/05-add-roles-to-db.yaml [hunk 9]
```diff
@@ -8,10 +8,10 @@ databaseChangeLog:
             columns:
               - column:
                   name: name
-                  value: ROLE_ADMIN
+                  value: ADMIN
```
Comment (NazarSmith228) [inline src/main/resources/db/changelog/changes/05-add-roles-to-db.yaml:11]: I'm not sure whether it is a good idea to update existing Liquibase migrations

please be careful with that and read [this](https://www.liquibase.com/blog/dealing-with-changing-changesets)

I would suggest to create a new changeSet, where you would update the existing DB records

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/mmosii/bookstore/controller/BookController.java (hunk 1 above)
Comment (mmosii) [reply src/main/java/mmosii/bookstore/controller/BookController.java:37]: Yeah, tried to change role names to shorter ones, but with hasRole it don't work, had to use hasAuthority.

File: src/main/resources/db/changelog/changes/05-add-roles-to-db.yaml (hunk 9 above)
Comment (mmosii) [reply src/main/resources/db/changelog/changes/05-add-roles-to-db.yaml:11]: I understand, just catching possibility to drop tables, while DB is empty, trying to not make my changelog a mess, before i even finish it =)

File: src/main/java/mmosii/bookstore/service/impl/BookServiceImpl.java [hunk 10]
```diff
@@ -63,9 +72,23 @@ public List<BookDto> search(BookSearchParametersDto searchParameters) {
...
     private void checkIfBookExistsById(Long id) {
         if (!bookRepository.existsById(id)) {
             throw new EntityNotFoundException("Can't find book with id " + id);
         }
     }
+
+    private void setCategories(CreateBookRequestDto requestDto, Book book) {
+        Set<Category> categorySet = requestDto.categoryIds().stream()
+                .map(categoryRepository::getReferenceById)
+                .collect(Collectors.toSet());
```
Comment (NazarSmith228) [inline src/main/java/mmosii/bookstore/service/impl/BookServiceImpl.java:93]: does it work? get reference returns a proxy with just id

when you save it to set it invokes hashCode, which in turn has to cause LazyInitException

does it really work for you?

**Review by NazarSmith228 — APPROVED** [review]
LGTM,  but please pay attention to my comment

File: src/main/java/mmosii/bookstore/service/impl/BookServiceImpl.java (hunk 10 above)
Comment (mmosii) [reply src/main/java/mmosii/bookstore/service/impl/BookServiceImpl.java:93]: worked, had open-in-view = true)
